### DISCUSSION TITLE
Timelapse and motion with esp32cam (and s3)

### DISCUSSION THREAD

**Discussion by btsimonh** [arendst/Tasmota#21468] [Show and tell]
Hi All,

I had time to update my repo about timelapse and motion capture using TAS here:

https://github.com/btsimonh/TasmotaBerryPLay/tree/master/timelapse

If you give it a go, please leave feedback, raise an issue there, or offer PRs....

(this is based on webcamv2 with motion detection, and also uses direct http serving off TAS's SDcard, OR posts images to a node-red http server)

br,

Simon

**Comment by btsimonh** [comment]
I made some improvements yesterday - putting the images into a folder structure which ensures no folder contains more than 10 files or folders.  This has proved valuable as with 1000 images, retrieving images close to the end was taking >5s.  Now regardless of where the image is in the sequence, the retrieval time is <500ms from SD.
This also vastly increase the speed of download of the whole folder.
Over 13 hours, my humble esp32cam (ai thinker style) has stored >9000 images, one every 10s.
It's not totally stable - suggest to close the TAS web interface whilst viewing or downloading.  Reading files from SD seems to stall after file open sometimes, hanging the main loop.

However, on the esp32S3cam board, I've not experienced any hangs yet, and the download is more than twice as quick. (~100ms/image).

**Reply by btsimonh** [reply]
and now a PR which makes it much more reliable on 'normal' esp32cam, and some updates to the berry code.

**Comment by btsimonh** [comment]
With the new enhancements, and the small reliability updates in latest dev, I had three devices recording images every 500ms last night.  2 x ai-thinker and 1x S3.  An AI -thinker recording larger images recorded ~43000 (I think storing the larger image took > 500ms sometimes), the other two ~65000 images.

The timelapse viewer performed fast and well in seeking to the images and replaying them.  I won't try to download them, as the download creates a zip in memory in browser...   but now to set them up for 1 hour timelapse and simultaneous motion detect (to a different folder) to watch my veggies grow this year :).

**Comment by btsimonh** [comment]
repo at https://github.com/btsimonh/TasmotaBerryPLay/tree/master/timelapse has been enhanced to make images zip download much more robust (my TAS is >50m from my router, and my laptop is less than stable on the wifi), and also to clarify image zip to .mp4 creation.

So far, my garden timelapse has recorded 505 images over ~4 weeks (hourly, excluding 'dark' hours).

Example .mp4 - watch TAS fall off in the first day or so, and the artichoke grow on the right:

https://github.com/arendst/Tasmota/assets/3137332/5126aa19-e3c6-4ed7-b36c-dcb3a207de2d

**Comment by rmschaller** [comment]
This project is of great interest to.  Running it on an ESP32-CAM board, I can build tasmota32-webcam successfully but when I add in the following, per the readme.md:

#define USE_WEBCAM_V2
#define USE_WEBCAM_MOTION
#define UFILESYS_STATIC_SERVING
#define ENABLE_RTSPSERVER
#define USE_SPI
#define USE_SDCARD
// problem should be resolved
//#undef USE_ESP32_WDT // if this problem still exists... else it WILL randomly crash.

in the user_config_override.h I get an error flagging line 935 of tasmota/tasmota_xdrv_driver/xdrv_50_filesystem.ino.  I wasn't sure about the line directly following the "problem should be resolved" comment.  Tried building with final line commented out and un-commented. Result was the same either way.  Any insight to this error would be greatly appreciated. Thanks

**Reply by btsimonh** [reply]
try without 
#define UFILESYS_STATIC_SERVING
if it all works except the fileserving, then we'll need to repair static fileserving, if required.

Most likely, the 'StaticRequestHandler' class has changed, and the function no longer matches the prototype it's overloading? - it may be a simple fix, or it may be absolute hell! :).

**Reply by rmschaller** [reply]
Thanks for the suggestion, removing #define UFILESYS_STATIC_SERVING resulted in a successful build.  So I wonder how this will affect the operation of the project  Based on the naming, without doing much research, this has something to do with the SD card.  Thanks

**Reply by joba-1** [reply]
why don’t you just show the build output instead of let people guess what happens?

**Reply by sfromis** [reply]
Having `UFILESYS_STATIC_SERVING` only matters if you use the command `UfsServe` to setup a web server.

**Reply by rmschaller** [reply]
`Compiling .pio/build/tasmota32-webcam/src/tasmota.ino.cpp.o
/Users/rschaller/Documents/PlatformIO/Projects/Tasmota-master/tasmota/tasmota_xdrv_driver/xdrv_50_filesystem.ino:935:10: error: 'bool StaticRequestHandlerAuth::handle(WebServer&, HTTPMethod, String)' marked 'override', but does not override
  935 |     bool handle(WebServer& server, HTTPMethod requestMethod, String requestUri) override {
      |          ^~~~~~
Retrieved `.pio/build/tasmota32-webcam/lib03a/WiFiHelper/WiFiHelper_ESP32.cpp.o' from cache
Retrieved `.pio/build/tasmota32-webcam/lib78b/tls_mini/StackThunk_light.cpp.o' from cache
*** [.pio/build/tasmota32-webcam/src/tasmota.ino.cpp.o] Error 1`

**Reply by rmschaller** [reply]
I just posted the original error before I got it to build per @btsimonh suggestion. Right now, I want to try out the basic functionality of the project so the web server isn't critical. - Thanks

**Reply by rmschaller** [reply]
@btsimonh I updloaded the built firmware files to the target but nothing seems to be happening. From what i've read ,once running, the inserted SD card will be formatted with directory structure but nothing was changed on the SD card.  Even a press of the reset button doesn't flash the onboard LED that would indicate a reset has been performed.  Just to confirm, my target board is an esp32-cam.  The metal shield covering the SoC device is stamped ESP32S.  I believe this is supported.  Is there some serial port debugging that I can monitor during the startup? If so, what baud rate?  I know my device isn't bricked as I can reflash another project and have it functioning as normal.  Here is the flashing of this project on my device, nothing looks suspicious - Thanks

`AVAILABLE: cmsis-dap, esp-bridge, esp-prog, espota, esptool, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa
CURRENT: upload_protocol = esptool
Looking for upload port...
Using manually specified: /dev/cu.usbserial-1410
Uploading .pio/build/tasmota32-webcam/firmware.bin
esptool.py v4.8.4
Serial port /dev/cu.usbserial-1410
Connecting.....
Chip is ESP32-D0WDQ6 (revision v1.0)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: c8:f0:9e:9d:87:c4
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Configuring flash size...
Auto-detected Flash size: 4MB
Flash will be erased from 0x00001000 to 0x00005fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x0000e000 to 0x0000ffff...
Flash will be erased from 0x00010000 to 0x000d1fff...
Flash will be erased from 0x000e0000 to 0x00227fff...
SHA digest in image updated
Compressed 16480 bytes to 11245...
Writing at 0x00001000... (100 %)
Wrote 16480 bytes (11245 compressed) at 0x00001000 in 0.5 seconds (effective 251.1 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 134...
Writing at 0x00008000... (100 %)
Wrote 3072 bytes (134 compressed) at 0x00008000 in 0.1 seconds (effective 465.3 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Writing at 0x0000e000... (100 %)
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 687.5 kbit/s)...
Hash of data verified.
Compressed 794464 bytes to 561095...
Writing at 0x00010000... (2 %)
Writing at 0x000168a0... (5 %)
Writing at 0x0001f03a... (8 %)
Writing at 0x0002512a... (11 %)
Writing at 0x0002a555... (14 %)
Writing at 0x0002fbac... (17 %)
Writing at 0x00035a06... (20 %)
Writing at 0x0003b23c... (22 %)
Writing at 0x00040576... (25 %)
Writing at 0x00045c09... (28 %)
Writing at 0x0004b270... (31 %)
Writing at 0x00050ae3... (34 %)
Writing at 0x00055e18... (37 %)
Writing at 0x0005b1e9... (40 %)
Writing at 0x00060830... (42 %)
Writing at 0x00065c2e... (45 %)
Writing at 0x0006b4da... (48 %)
Writing at 0x000712eb... (51 %)
Writing at 0x000770d2... (54 %)
Writing at 0x0007c149... (57 %)
Writing at 0x000814b5... (60 %)
Writing at 0x00086702... (62 %)
Writing at 0x0008bd21... (65 %)
Writing at 0x00091339... (68 %)
Writing at 0x000969a0... (71 %)
Writing at 0x0009bda5... (74 %)
Writing at 0x000a1499... (77 %)
Writing at 0x000a73d3... (80 %)
Writing at 0x000ace0f... (82 %)
Writing at 0x000b23ec... (85 %)
Writing at 0x000b7f00... (88 %)
Writing at 0x000bcf90... (91 %)
Writing at 0x000c4aa5... (94 %)
Writing at 0x000ca36f... (97 %)
Writing at 0x000d0570... (100 %)
Wrote 794464 bytes (561095 compressed) at 0x00010000 in 14.7 seconds (effective 431.0 kbit/s)...
Hash of data verified.
Compressed 1339920 bytes to 898889...
Writing at 0x000e0000... (1 %)
Writing at 0x000e6d18... (3 %)
Writing at 0x000ed749... (5 %)
Writing at 0x000f6f31... (7 %)
Writing at 0x000ff4ff... (9 %)
Writing at 0x00108566... (10 %)
Writing at 0x00110ac7... (12 %)
Writing at 0x0011afed... (14 %)
Writing at 0x00120a6c... (16 %)
Writing at 0x00126499... (18 %)
Writing at 0x0012b799... (20 %)
Writing at 0x00130c9d... (21 %)
Writing at 0x00136540... (23 %)
Writing at 0x0013bc13... (25 %)
Writing at 0x001416fa... (27 %)
Writing at 0x00147640... (29 %)
Writing at 0x0014dc2e... (30 %)
Writing at 0x00153af9... (32 %)
Writing at 0x001592de... (34 %)
Writing at 0x0015edcc... (36 %)
Writing at 0x00164d70... (38 %)
Writing at 0x0016ac52... (40 %)
Writing at 0x00170b2c... (41 %)
Writing at 0x00176f48... (43 %)
Writing at 0x0017ccb0... (45 %)
Writing at 0x001823a6... (47 %)
Writing at 0x00187cf0... (49 %)
Writing at 0x0018d790... (50 %)
Writing at 0x001931ba... (52 %)
Writing at 0x00198d2e... (54 %)
Writing at 0x0019e05e... (56 %)
Writing at 0x001a389c... (58 %)
Writing at 0x001a8efc... (60 %)
Writing at 0x001ae955... (61 %)
Writing at 0x001b45b6... (63 %)
Writing at 0x001b9d6e... (65 %)
Writing at 0x001bf2db... (67 %)
Writing at 0x001c51c5... (69 %)
Writing at 0x001ca84e... (70 %)
Writing at 0x001cf9be... (72 %)
Writing at 0x001d4d7a... (74 %)
Writing at 0x001d9f79... (76 %)
Writing at 0x001df67b... (78 %)
Writing at 0x001e4e3e... (80 %)
Writing at 0x001ea214... (81 %)
Writing at 0x001ef645... (83 %)
Writing at 0x001f4e25... (85 %)
Writing at 0x001facec... (87 %)
Writing at 0x0020086e... (89 %)
Writing at 0x00205efd... (90 %)
Writing at 0x0020bc49... (92 %)
Writing at 0x00210ce0... (94 %)
Writing at 0x0021664e... (96 %)
Writing at 0x0021c2df... (98 %)
Writing at 0x00221a88... (100 %)
Wrote 1339920 bytes (898889 compressed) at 0x000e0000 in 23.7 seconds (effective 452.3 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
==================================================================================== [SUCCESS]`

**Reply by sfromis** [reply]
You can monitor serial output with the same wiring as when flashing. Bit rate 115200, 8N1

**Reply by rmschaller** [reply]
OK, perhaps I wasn't patient enough for the processor speed :-) and I see debug info coming from the serial port.  It is working now as the web interface is showing the cam video as expected. Again, thanks for the support!
